kedro/config/omegaconf_config.py

```diff
@@ -0,0 +1,258 @@
+            sorted_keys = ", ".join(sorted(common_keys))
+            msg = (
+                "Config from path '%s' will override the following "
+                "existing top-level config keys: %s"
+            )
+            _config_logger.debug(msg, env_path, sorted_keys)
+
+        config.update(env_config)
+
+        if not config:
+            raise MissingConfigException(
+                f"No files of YAML or JSON format found in {base_path} or {env_path} matching"
+                f" the glob pattern(s): {[*self.config_patterns[key]]}"
+            )
+        return config
+
+    def __repr__(self):  # pragma: no cover
+        return (
+            f"OmegaConfLoader(conf_source={self.conf_source}, env={self.env}, "
+            f"config_patterns={self.config_patterns})"
+        )
+
+    def load_and_merge_dir_config(self, conf_path: str, patterns: Iterable[str]):
+        """Recursively load and merge all configuration files in a directory using OmegaConf,
+        which satisfy a given list of glob patterns from a specific path.
+
+        Args:
+            conf_path: Path to configuration directory.
+            patterns: List of glob patterns to match the filenames against.
+
+        Raises:
+            MissingConfigException: If configuration path doesn't exist or isn't valid.
+            ValueError: If two or more configuration files contain the same key(s).
+            ParserError: If config file contains invalid YAML or JSON syntax.
+
+        Returns:
+            Resulting configuration dictionary.
+
+        """
+        if not Path(conf_path).is_dir():
+            raise MissingConfigException(
+                f"Given configuration path either does not exist "
+                f"or is not a valid directory: {conf_path}"
+            )
+
+        paths = [
+            Path(each).resolve()
+            for pattern in patterns
+            for each in iglob(f"{str(conf_path)}/{pattern}", recursive=True)
+        ]
+        deduplicated_paths = set(paths)
+        config_files_filtered = [
+            path for path in deduplicated_paths if self._is_valid_config_path(path)
+        ]
+
+        config = {}
+        aggregate_config = []
+        seen_file_to_keys = {}  # type: Dict[Path, AbstractSet[str]]
+
+        for config_filepath in config_files_filtered:
+            try:
+                single_config = OmegaConf.load(config_filepath)
+                config = single_config
+            except (ParserError, ScannerError) as exc:
+                assert exc.problem_mark is not None
+                line = exc.problem_mark.line
+                cursor = exc.problem_mark.column
+                raise ParserError(
+                    f"Invalid YAML or JSON file {config_filepath}, unable to read line {line}, "
+                    f"position {cursor}."
+                ) from exc
+
+            _check_duplicate_keys(
+                seen_file_to_keys, config_filepath, dict(single_config)
+            )
```

Comment:
Could we do this check only at the end of the loop, so we can give error message for all duplicates at once?

In fact, maybe we could keep only one variable, say `config_per_file`, mapping `filepath -> config`, and then check the duplicates in a separate loop, which will produce all errors at once.

After all `aggregate_config` is just `config_per_file.values()` and `seen_file_to_keys = {file: set(config.keys()) for file, config in config_per_file.items()}`. Updating two data structures in the same loop is harder to read.

Reply:
So instead you would load all the files and then compare all of them against each other to find duplicates? Wouldn't that me a lot more inefficient? I guess the benefit is that you would immediately know all duplicates, but the downside is that if you have a lot of config everything needs to be loaded first and then checked before you know something is wrong.

Reply:
I've rewritten this to now check duplicates between all config files in a directory + subdirectories.